### DISCUSSION TITLE
Android programmatic indirect lighting possible?

### DISCUSSION THREAD

**Discussion by srMarlins** [google/filament#5219] [Q&A] [answered]
Is it possible to build indirect lighting programmatically on Android?

I'm looking to render with a transparent background (no skybox?) without an ibl/ktx file (to minimize apk size). We have to match web/android so I was going to use [modelviewer's neutral environment](https://github.com/google/model-viewer/pull/2916) but it's ~1MB. 

Right now I'm just using the following as placeholder values which renders most models correctly but [khronos box interleaved](https://modelviewer.dev/fidelity/#khronos-BoxInterleaved) doesn't have the sides correctly illuminated (currently just black). 
```
IndirectLight.Builder()
            .intensity(20000f)
            .radiance(1, floatArrayOf(1f, 1f, 1f))
            .irradiance(1, floatArrayOf(1f, 1f, 1f))
            .build(engine)
```

For reference the only source of direct lighting within the scene is 
```
val (r, g, b) = cct(8_000.0f)
        LightManager.Builder(Type.DIRECTIONAL)
            .castLight(true)
            .direction(0.0f, -0.75f, -1.25f)
            .castShadows(true)
            .intensity(110000f)
            .color(r, g, b)
            .build(engine, light)
```

if there's a better way of doing this by generating a small (max 300 KB) environment file I'd also do that :)

**Comment by romainguy** [comment, marked as answer]
You can generate a smaller IBL using that neutral environment:

```
cmgen -f ktx -s 64 -x /Users/romainguy/Desktop/test ~/Downloads/neutral.hdr
```

By limiting the size to 64x64 you'll get a .ktx that's ~130 KiB uncompressed and ~45 KiB once zipped in an APK. You could also try -s 128 and see how well the result compressed (it should compress extremely well given this IBL).
